ghc-toolkit/boot-libs/rts/PrimOps.cmm

```diff
@@ -1432,31 +1431,109 @@ stg_newMVarzh ()
     StgStack_sp(stack) = sp;                    \
     lval = W_[sp - WDS(1)];
 
-
 stg_takeMVarzh ( P_ mvar /* :: MVar a */ )
 {
     W_ val, info, tso, q;
 
+    LOCK_CLOSURE(mvar, info);
+
     /* If the MVar is empty, put ourselves on its blocking queue,
      * and wait until we're woken up.
      */
     if (StgMVar_value(mvar) == stg_END_TSO_QUEUE_closure) {
-        jump stg_raisezh(base_GHCziIOziException_blockedIndefinitelyOnMVar_closure);
+        if (info == stg_MVAR_CLEAN_info) {
+            ccall dirty_MVAR(BaseReg "ptr", mvar "ptr");
+        }
+
+        // We want to put the heap check down here in the slow path,
+        // but be careful to unlock the closure before returning to
+        // the RTS if the check fails.
+        ALLOC_PRIM_WITH_CUSTOM_FAILURE
+            (SIZEOF_StgMVarTSOQueue,
+             unlockClosure(mvar, stg_MVAR_DIRTY_info);
+             GC_PRIM_P(stg_takeMVarzh, mvar));
+
+        q = Hp - SIZEOF_StgMVarTSOQueue + WDS(1);
+
+        SET_HDR(q, stg_MVAR_TSO_QUEUE_info, CCS_SYSTEM);
+        StgMVarTSOQueue_link(q) = END_TSO_QUEUE;
+        StgMVarTSOQueue_tso(q)  = CurrentTSO;
+
+        if (StgMVar_head(mvar) == stg_END_TSO_QUEUE_closure) {
+            StgMVar_head(mvar) = q;
+        } else {
+            StgMVarTSOQueue_link(StgMVar_tail(mvar)) = q;
+            ccall recordClosureMutated(MyCapability() "ptr",
+                                             StgMVar_tail(mvar));
+        }
+        StgTSO__link(CurrentTSO)       = q;
```

Comment:
In our `StgTSO` object scavenge logic, did we handle the `__link` field as well?

Reply:
Currently the GC evacuate all the TSOs (even the completed ones). I have updated #310 so that we don't forget to fix this.

Reply:
> evacuates all the TSOs
I think you really mean "scavenge" here? (Traversing all fields of an object), since TSOs are all pinned for now, we don't really evacuate them

Reply:
Well we call "evacuateClosure" on them (the fact that they don't really move is an implementation detail IMO). But what I meant was that the `link` field contains a list of TSOs and we evacuate them all anyway, so it doesn't change anything if we consider this field or not during the TSO scavenging (for now).